File: code/python/algorithm_thinking_article-15.py
"""Shaobing sorting algorithm"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().__init__(*args, **kwargs)


class _Solution:
    def __init__(self):
        self.res = []
    def pancakeSort(self, cakes: List[int]) -> List[int]:
        self.sort(cakes, len(cakes))
        return self.res
    def sort(self, cakes: List[int], n: int) -> None:
        if n == 1:
            return
        maxCake = 0
        maxCakeIndex = 0
        for i in range(n):
            if cakes[i] > maxCake:
                maxCakeIndex = i
                maxCake = cakes[i]
        self.reverse(cakes, 0, maxCakeIndex)
        self.res.append(maxCakeIndex + 1)
        self.reverse(cakes, 0, n - 1)
        self.res.append(n)
        self.sort(cakes, n - 1)
    def reverse(self, arr: List[int], i: int, j: int) -> None:
        while i < j:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1

File: code/javascript/algorithm_thinking_article-15.js
/**
 * Shaobing sorting algorithm
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return sort(...args);
}

var pancakeSort = function(cakes) {
    const res = [];
    sort(cakes, cakes.length);
    return res;
    function sort(cakes, n) {
        if (n == 1) return;
        let maxCake = 0;
        let maxCakeIndex = 0;
        for (let i = 0; i < n; i++) {
            if (cakes[i] > maxCake) {
                maxCakeIndex = i;
                maxCake = cakes[i];
            }
        }
        reverse(cakes, 0, maxCakeIndex);
        res.push(maxCakeIndex + 1);
        reverse(cakes, 0, n - 1);
        res.push(n);
        sort(cakes, n - 1);
    }
    function reverse(arr, i, j) {
        while (i < j) {
            let temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
            j--;
        }
    }
};
